habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -386,6 +386,290 @@ def step(self, *args, is_last_action, **kwargs):
+        (w - v * np.sin(phi) / D) / phi > v * np.cos(phi) / D
+        v < (w / phi) / (np.sin(phi) / D / phi + np.cos(phi) / D)
+        v < w * D / (np.sin(phi) + phi * np.cos(phi))
+
+        (D = linear error, phi = angular error)
+        """
+        assert lin_err >= 0.0
+        assert heading_diff >= 0.0
+
+        max_heading_ang = np.pi / 10
+        if heading_diff > max_heading_ang:
+            return 0.0
+        else:
+            return (
+                w_max
+                * lin_err
+                / (
+                    np.sin(heading_diff)
+                    + heading_diff * np.cos(heading_diff)
+                    + 1e-5
+                )
+            )
+
+    def _feedback_simple(self, base_pose_err):
+        """
+        Feedback controllers
+        """
+        v_cmd = w_cmd = 0
+
+        lin_err_abs = np.linalg.norm(base_pose_err[0:2])
+        ang_err = base_pose_err[2]
+
+        # Go to goal XY position if not there yet
+        if lin_err_abs > self.lin_error_tol:
+            heading_err = np.arctan2(base_pose_err[1], base_pose_err[0])
+            heading_err_abs = abs(heading_err)
+
+            if abs(heading_err_abs - np.pi) <= 0.01:
+                heading_err_abs = 0
+                heading_err = 0
+
+            # Compute linear velocity and allow the agent to move backward
+            v_raw = self._velocity_feedback_control(
+                np.sign(base_pose_err[0])*lin_err_abs, self.acc_lin, self.v_max
+            )
+            v_limit = self._turn_rate_limit(
+                lin_err_abs,
+                heading_err_abs,
+                self.w_max / 2.0,
+                tol=self.lin_error_tol,
+            )
+            # Allow the agent to move backward
+            v_cmd = np.clip(v_raw, -v_limit, v_limit)
+
+            # Compute angular velocity
+            w_cmd = self._velocity_feedback_control(
+                heading_err, self.acc_ang, self.w_max
+            )
+
+        # Rotate to correct yaw if yaw tracking is on and XY position is at goal
+        elif abs(ang_err) > self.ang_error_tol and self.track_yaw:
+            # Compute angular velocity
+            w_cmd = self._velocity_feedback_control(
+                ang_err, self.acc_ang, self.w_max
+            )
+
+        return v_cmd, w_cmd
+
+    def _transform_global_to_base(self, base_pose, current_pose, sim=None):
+        """
+        Transforms the point cloud into geocentric frame to account for
+        camera position
+        Input:
+            base_pose                     : target goal ...x3
+            current_pose            : base position (x, y, theta (radians))
+        Output:
+            base_pose : ...x3
+        """
+
+        trans = sim.robot.base_transformation
+        goal_pos = trans.inverted().transform_point(
+            np.array([base_pose[0], sim.robot.base_pos[1], base_pose[1]])
+        )
+
+        error_t = base_pose[2] - current_pose[2]
```

Comment:
this is the theta error, can you actually subtract two angles like this? what if they are like 0 and np.pi?

Reply:
I fixed this, thanks!